src/components/PreventNavigationModal.tsx

```diff
@@ -0,0 +1,127 @@
+/**
+ * Invoice Ninja (https://invoiceninja.com).
+ *
+ * @link https://github.com/invoiceninja/invoiceninja source repository
+ *
+ * @copyright Copyright (c) 2022. Invoice Ninja LLC (https://invoiceninja.com)
+ *
+ * @license https://www.elastic.co/licensing/elastic-license
+ */
+
+import { useTranslation } from 'react-i18next';
+import { Modal } from './Modal';
+import { Button } from './forms';
+import { useAtom, useAtomValue } from 'jotai';
+import { useEffect, useState } from 'react';
+import {
+  blockedNavigationActionAtom,
+  isNavigationModalVisibleAtom,
+} from '$app/common/hooks/usePreventNavigation';
+import { lastHistoryLocationAtom, preventLeavingPageAtom } from '$app/App';
+import { useLocation, useNavigate } from 'react-router-dom';
+
+export function PreventNavigationModal() {
+  const [t] = useTranslation();
+  const navigate = useNavigate();
+  const location = useLocation();
+
+  const [lastHistoryLocation, setLastHistoryLocation] = useAtom(
+    lastHistoryLocationAtom
+  );
+  const blockedNavigationAction = useAtomValue(blockedNavigationActionAtom);
+  const [preventLeavingPage, setPreventLeavingPage] = useAtom(
+    preventLeavingPageAtom
+  );
+  const [isNavigationModalVisible, setIsNavigationModalVisible] = useAtom(
+    isNavigationModalVisibleAtom
+  );
+
+  const [isModalVisible, setIsModalVisible] = useState<boolean>(false);
+
+  const { nonPreventedLocations } = lastHistoryLocation;
+
+  const handleDiscardChanges = () => {
+    const isBrowserBackAction = preventLeavingPage.actionKey === 'browserBack';
+
+    const { url, externalLink, fn } = blockedNavigationAction || {};
+
+    if (!externalLink) {
+      setLastHistoryLocation((current) => ({
+        ...current,
+        lastLocation: '',
+      }));
+      setPreventLeavingPage({ prevent: false, actionKey: undefined });
+    }
+
+    setIsNavigationModalVisible(false);
+
+    const numberOfNonPreventedLocations = nonPreventedLocations.length;
+
+    let lastNonPreventedLocation =
+      nonPreventedLocations[numberOfNonPreventedLocations - 1];
+
+    lastNonPreventedLocation =
+      lastNonPreventedLocation !== location.pathname
+        ? lastNonPreventedLocation
+        : nonPreventedLocations[numberOfNonPreventedLocations - 2];
+
+    if (isBrowserBackAction && lastNonPreventedLocation) {
+      navigate(lastNonPreventedLocation);
+    }
+
+    if (blockedNavigationAction) {
+      if (url) {
+        if (url === 'back') {
+          lastNonPreventedLocation && navigate(lastNonPreventedLocation);
+        } else {
+          if (externalLink) {
+            window.open(url, '_blank');
+          } else {
+            navigate(url);
+          }
+        }
+      }
+
+      fn?.();
+    }
+  };
```

Comment:
This looks fairly complex. Any chance we can provide some tests for this one? We definitely don't want navigation just randomly failing, that'd be bad.

Reply:
@beganovich Sure, explaining the entire logic in this comment: https://github.com/invoiceninja/ui/pull/1419#pullrequestreview-1790567160